rgw: beast frontend reworks pause/stop and yields during body io

cbodley opened the pull request:
the beast frontend's ClientIO impl uses async_read/write on the socket to yield its coroutine. this means that process_request() is no longer fully synchronous

but the frontend's use of io_context.stop() for pause() and stop() relied on process_request() being treated as a single unit of 'work' to prevent it from stopping in the middle of a request. this motivated the changes to pause and stop to make more natural use of the io_context (and, as a side effect, makes it easier for other async work to be scheduled on this io_context without interfering with pause)

a new async::SharedMutex is used to implement pause (analagous to civetweb's use of RWLock). this SharedMutex presents a unified interface for sync and async waits. the async waits use a new generic async::Completion primitive

when these Completions are used with a boost::asio::strand executor (as is the case with boost's stackful coroutines), the associated allocator is a 'recycling_allocator' that reuses memory from the previous async operation. so in effect, neither the sync nor async waits in SharedMutex require heap allocation

cbodley commented:
no related failures in http://pulpito.ceph.com/cbodley-2018-04-06_12:52:02-rgw-wip-rgw-beast-async-distro-basic-smithi/

chardan commented on src/common/async/bind_handler.h:
```diff
@@ -0,0 +1,111 @@
+#include <tuple>
+#include <boost/asio.hpp>
+
+namespace ceph::async {
```
Tiny (very tiny) reservation about this. If someone writes:
using namespace ceph; 
...and-- as is unfortunately frequently-- someone's imported namespace std, we'll have a collision.

adamemerson replied:
We'll be destroying that soon. Ish.

chardan commented on src/common/async/bind_handler.h:
```diff
@@ -0,0 +1,111 @@
+      args(std::move(args))
+  {}
+
+  void operator()() & {
```
Great that you have these bases covered! This enables CompletionHandler objects to be called in a variety of contexts and do the right thing with their handler and args. Yay!

cbodley replied:
thanks for the reminder - i've actually gone back and forth on this one, and could use some advice

both asio and beast have a similar mechanism (`boost::asio::detail::bind_handler()` and `boost::beast::bind_handler()`), but without the `&&` overload. and there's still a missing piece that makes this `&&` overload useful, because executors will always invoke handlers as an lvalue reference

my first intuition was to implement this `operator() &` overload with `std::move(args)` as a workaround, but that would be destructive to the arguments, and you wouldn't be able to use this class as a general _reusable_ function object

to actually make use of this `&&` overload with executors, you would need an extra wrapper that converts the `operator() &` into `operator() &&`, which would look something like `forward_handler(bind_handler(h, args...))`

so `forward_handler()` would be a requirement for async operations with a move-only type in their Signature, i.e. `void(error_code, unique_ptr<T>)` - and `async::Completion` would need to do that also because it's generic over all Signatures

that would allow `SharedMutex::async_lock_shared()` to use a `Signature = void(error_code, std::shared_lock)` for a more natural interface:
```diff
-    pause_mutex.async_lock_shared(yield[ec]);
+    auto lock = pause_mutex.async_lock_shared(yield[ec]);
     if (ec) {
       return;
     }
-    std::shared_lock lock{pause_mutex, std::adopt_lock};
```

the question, though, is whether this extra layer of wrapping with `forward_handler()` is actually worth it, or if this `operator() &` overload really should just `std::move(args)`? i'm leaning towards the latter now, because i don't think this `CompletionHandler` is all that useful outside of the fact that it preserves the associated allocator/executor - `std::bind()` is better in every other way (supports partial binding, placeholders, etc)

cbodley replied:
thanks, that helps. i'm happy with the composition of forward_handler and bind_handler here. both can be optimized out anyway (see https://godbolt.org/g/oeokp8), so it doesn't cost us anything

chardan commented on src/common/async/shared_mutex.h:
```diff
@@ -0,0 +1,210 @@
+{
+  try {
+    impl.cancel();
+  } catch (...) {
```
It might be worth commenting on this. I'm assuming you want to swallow any exceptions to avoid throwing from the dtor, but OTOH taking no action might deserve a comment.

cbodley commented:
rebased

adamemerson commented:
Jenkins, retest docs please.

adamemerson commented:
Fine.

Jenkins, retest this you jerk.

cbodley commented:
@adamemerson thanks for the review. i'm conflicted about merging this now, since there's a slight regression in the ability to trace a given request through radosgw logs over multiple threads. https://github.com/ceph/ceph/pull/21700 proposes a framework for this, but will need more work to get full coverage. i'd prefer to pull the common/async bits into a separate pr

cbodley commented on src/common/async/shared_mutex.h:
```diff
@@ -0,0 +1,428 @@
+  /// wait synchronously for an exclusive lock. if an error occurs before the
+  /// lock is granted, that error is assigned to 'ec'
+  void lock(boost::system::error_code& ec);
+
```
oh cool, i hadn't seen that pattern before! do you know if the standard library provides anything similar? if not, i think i'll leave it as is to simplify the eventual conversion from boost::asio -> std::net

adamemerson replied:
It does not, unfortunately. (Maybe they'll add something by the time C++20 comes along, but I haven't heard about it.)

adamemerson commented:
@cbodley If you could pull out the common code into a separate PR that we could fast track in, I'd appreciate it.

cbodley commented:
rebased following merge of SharedMutex

cbodley commented:
this has been blocked on log refactoring work, so that we can trace requests that suspend and resume on a different thread. i've removed the commit `rgw: beast frontend uses yield_context to read/write body` and pushed that to a separate branch https://github.com/cbodley/ceph/commits/wip-rgw-beast-async-body. i rebased the other frontend changes here so we can move forward to testing/merging

cbodley commented:
passed in http://pulpito.ceph.com/cbodley-2018-09-07_19:00:38-rgw-wip-rgw-beast-async-distro-basic-smithi/ with --filter beast
